clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -211,6 +213,15 @@ private static FilterResult filterTo(TopicPartition partition, Iterable<MutableR
                     maxTimestamp = info.maxTimestamp;
                     shallowOffsetOfMaxTimestamp = info.shallowOffsetOfMaxTimestamp;
                 }
+            } else if (batchRetention == BatchRetention.RETAIN_EMPTY) {
+                if (batchMagic < RecordBatch.MAGIC_VALUE_V2)
+                    throw new IllegalStateException("Empty batches are only supported for magic v2 and above");
+
+                bufferOutputStream.maybeExpandBuffer(DefaultRecordBatch.RECORD_BATCH_OVERHEAD);
+                DefaultRecordBatch.writeEmptyHeader(bufferOutputStream.buffer(), batchMagic, batch.producerId(),
```

Comment:
That's a good question. Unlike the base offset, we currently keep the base timestamp in sync with the first message. If there are no messages left, it's not clear what, if anything, should be used. Maybe NO_TIMESTAMP would be more consistent? The same goes for the max timestamp in fact: if there are no messages left, then maybe the max timestamp should be NO_TIMESTAMP?

In retrospect, it might have been more consistent to preserve the base timestamp and max timestamp from the original batch as well. It would have made this case more straightforward anyway.